scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/resources/StateQueryParam.java

```diff
@@ -1,28 +1,61 @@
 package gov.nasa.jpl.aerie.scheduler.constraints.resources;
 
+import com.google.common.collect.Lists;
+import gov.nasa.jpl.aerie.constraints.model.DiscreteProfilePiece;
+import gov.nasa.jpl.aerie.constraints.model.LinearProfilePiece;
+import gov.nasa.jpl.aerie.constraints.model.SimulationResults;
 import gov.nasa.jpl.aerie.constraints.time.Window;
 import gov.nasa.jpl.aerie.merlin.protocol.types.SerializedValue;
 import gov.nasa.jpl.aerie.scheduler.model.Plan;
 import gov.nasa.jpl.aerie.scheduler.constraints.timeexpressions.TimeExpression;
 import gov.nasa.jpl.aerie.scheduler.constraints.activities.VariableArgumentComputer;
 
+import java.util.function.Predicate;
+
 /**
  * Class allowing to define state query expression for instantiation of parameters
  */
 public class StateQueryParam implements VariableArgumentComputer {
 
-  public final QueriableState state;
+  public final String nameResource;
   public final TimeExpression timeExpr;
 
-  public StateQueryParam(QueriableState state, TimeExpression timeExpression) {
-    this.state = state;
+  public StateQueryParam(String nameResource, TimeExpression timeExpression) {
+    this.nameResource = nameResource;
     this.timeExpr = timeExpression;
   }
 
-  public SerializedValue getValue(Plan plan, Window win) {
-    var time = timeExpr.computeTime(plan, win);
+  public SerializedValue getValue(SimulationResults simulationResults, Plan plan, Window win) {
+    var time = timeExpr.computeTime(simulationResults, plan, win);
     if (!time.isSingleton()) {
       throw new RuntimeException(" Time expression in StateQueryParam case must be singleton");
     }
-    return state.getValueAtTime(time.start);  }
+    final var queryT = time.start;
+
+    //TODO: unify necessary generic profile operations in Profile interface to avoid special casing
+    if (simulationResults.realProfiles.containsKey(this.nameResource)) {
+      //TODO: improve the profile data structure to allow fast time-keyed query
+      //TODO: improve Window to allow querying containment in interval directly
+      //for now we look for the last matching profile segment, if any (last to get latest if any overlaps)
+      final var profile = simulationResults.realProfiles.get(this.nameResource);
+      final Predicate<LinearProfilePiece> containsQueryTimeP = (piece)
+          -> !Window.intersect(piece.window, Window.at(queryT)).isEmpty();
+      final var matchPiece = Lists.reverse(profile.profilePieces).stream()
+                                  .filter(containsQueryTimeP).findFirst();
+      final var dblVal = matchPiece.map(piece -> piece.valueAt(queryT)).orElse(null);
+      return SerializedValue.of(dblVal);
```

Comment:
Rather than throw a null pointer exception (which is what will happen if we call `SerializedValue.of(null)`), we can throw an Error:

```suggestion
      final var piece = Lists.reverse(profile.profilePieces)
                             .stream()
                             .filter(containsQueryTimeP)
                             .findFirst()
                             .orElseThrow(() -> new Error(
                                 "Linear profile for %s not have a segment at the desired time %s"
                                     .formatted(nameResource, queryT)));
    return SerializedValue.of(piece.valueAt(queryT));
```

The other cases below should also probably throw errors - alternatively, this method could return an `Optional`, if it's not truly an error situation.